plugins/net_plugin/net_plugin.cpp

```diff
@@ -1238,67 +1301,128 @@ namespace eosio {
       return true;
    }
 
-   void connection::enqueue( const net_message& m ) {
-      verify_strand_in_this_thread( strand, __func__, __LINE__ );
-      go_away_reason close_after_send = no_reason;
-      if (std::holds_alternative<go_away_message>(m)) {
-         close_after_send = std::get<go_away_message>(m).reason;
+   //------------------------------------------------------------------------
+
+   using send_buffer_type = std::shared_ptr<std::vector<char>>;
+
+   struct buffer_factory {
+
+      /// caches result for subsequent calls, only provide same net_message instance for each invocation
+      const send_buffer_type& get_send_buffer( const net_message& m ) {
+         if( !send_buffer ) {
+            send_buffer = create_send_buffer( m );
+         }
+         return send_buffer;
       }
 
-      const uint32_t payload_size = fc::raw::pack_size( m );
+   protected:
+      send_buffer_type send_buffer;
 
-      const char* const header = reinterpret_cast<const char* const>(&payload_size); // avoid variable size encoding of uint32_t
-      constexpr size_t header_size = sizeof(payload_size);
-      static_assert( header_size == message_header_size, "invalid message_header_size" );
-      const size_t buffer_size = header_size + payload_size;
+   protected:
+      static send_buffer_type create_send_buffer( const net_message& m ) {
+         const uint32_t payload_size = fc::raw::pack_size( m );
 
-      auto send_buffer = std::make_shared<vector<char>>(buffer_size);
-      fc::datastream<char*> ds( send_buffer->data(), buffer_size);
-      ds.write( header, header_size );
-      fc::raw::pack( ds, m );
+         const char* const header = reinterpret_cast<const char* const>(&payload_size); // avoid variable size encoding of uint32_t
+         constexpr size_t header_size = sizeof(payload_size);
+         static_assert( header_size == message_header_size, "invalid message_header_size" );
```

Comment:
This static_assert is fairly useless. It carefully verifies that a uint32_t is 4 bytes. So, it complies with the C++11 standard? How helpful. If we really need to check for such a thing, we should put it in cmake.

Reply:
It is not verifying that uint32_t is 4 bytes so much as it is verifying that the `header_size` remains the same as `message_header_size` since it is used during the read. The idea is if you change this code you better also change all uses of `message_header_size`.

Reply:
payload_size is a uint32_t.  By definition (in the standard), uint32_t is a 4 byte integer.  So `sizeof(payload_size)` is 4, always.  constexpr size_t header_size == 4.  If the intent is to be sure message_header_size is always 4 bytes, that's a ridiculously convoluted way to achieve it.  Better to just change the definition: `constexpr auto message_header_size = sizeof(uint32_t);`, dispense with `header_size` here and just use `message_header_size` directly to write the bytes of `payload_size`.  That would be more consistent with how it's used during reading anyway.

Meanwhile `static_assert(
    sizeof(decltype(fc::raw::pack_size(std::declval<net_message&>())))==sizeof(uint32_t),
     "pack_size return value doesn't always fit in 4 bytes");` will fail.  (But we're not trying to solve that problem.  I just couldn't let go of the expression after I figured out the syntax...)

Reply:
Updated to use `message_header_size`